State dict pre hook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88541

Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287.

rohan-varma committed:
State dict pre hook

[ghstack-poisoned]

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

awgu commented on torch/nn/modules/module.py:
```diff
@@ -1585,6 +1589,16 @@ def _register_state_dict_hook(self, hook):
         return handle
 
+    def register_state_dict_pre_hook(self, hook):
+        r"""These hooks will be called with arguments: `self`, `prefix`,
```
nit: double backtick for code type in public docstrings

awgu commented on test/test_nn.py:
```diff
@@ -16147,6 +16147,60 @@ def my_post_load_hook(mod, _):
+        self.assertTrue(hasattr(m, '_state_dict_pre_hooks'))
+        delattr(m, '_state_dict_pre_hooks')
+        # Save and load, ensure we can still call state_dict
+        # without running issues.
```
nit:
```suggestion
        # without running into issues.
```

awgu commented on torch/nn/modules/module.py:
```diff
@@ -1585,6 +1589,16 @@ def _register_state_dict_hook(self, hook):
         self._state_dict_hooks[handle.id] = hook
         return handle
 
+    def register_state_dict_pre_hook(self, hook):
```
Curious: Do we want this to be public? If so, is there a reason why `_register_state_dict_hook()` is private?

albanD replied:
We do want to make these properly public. The issue linked in the description should give details about that!

awgu reviewed (commented):
LGTM, but I will let Core team to approve.

albanD reviewed (approved):
Sounds good to me beyond the nits from @awgu

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

rohan-varma committed:
Update on "State dict pre hook"


Implements the API discussed in https://github.com/pytorch/pytorch/issues/75287. 


[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1668,6 +1672,16 @@ def _register_state_dict_hook(self, hook):
 
+    def register_state_dict_pre_hook(self, hook):
+        r"""These hooks will be called with arguments: ``self``, ``prefix``,
+        and ``keep_vars`` before calling ``state_dict`` on ``self``. The registered
```
A comment from the post hook discussion is that since `state_dict()` is not actually called recursively, this hook will only be triggered on the top level Module.
We should clarify that in the doc if that is expected. Or think about updating the implementation/testing if it is not.
